Small molecule chromatogram library support

labkey-jeckels opened the pull request:
#### Rationale
We initially supported just proteomics (proteins and peptides) chromatogram libraries. We want to support small molecule use cases too.

#### Changes
* Rearrange web parts into separate portal tabs to reduce the number of large grids on one page
* Create a new version of .clib files with schema additions for small molecule data
* Report and resolve conflicts for small molecules too

Commit:
Create small molecule tables in chromatogram libraries, add a few new columns to protemics tables

Commit:
Checkpoint - write small molecule data to clib files

Commit:
Merge branch 'develop' into fb_smallMoleculeChromLib

Commit:
Import handling

Commit:
Resolve conflicts for small molecules

Remove moleculeList table from CLIB since we're not populating it

Commit:
Improve styling on download grid

labkey-jeckels committed:
Test files for automation

Commit:
Move grids to separate tables for molecules, peptides, and proteins, and show once we have enough data loaded to be informed

Commit:
Add automated test coverage

Fine tune the columns in grids other minor UI tweaks

vagisha commented on resources/queries/targetedms/librarymoleculeprecursor/LibraryPrecursors.qview.xml:
```diff
@@ -0,0 +1,28 @@
+    </filters>
+
+    <sorts>
+        <sort column="IonFormula" />
```
Can we sort on the MoleculeId/MoleculeName here instead of IonFormula and CustomIonName both of which can be null?

labkey-jeckels replied:
Added MoleculeName as primary sort, but left the others as secondary ordering

vagisha commented on src/org/labkey/targetedms/query/ConflictResultsManager.java:
```diff
@@ -88,67 +87,76 @@ private ConflictResultsManager() {}
+    {
+        List<ConflictPrecursor> result = new ArrayList<>();
+        result.addAll(getConflictedPrecursors(container, TargetedMSManager.getTableInfoPeptide(), TargetedMSManager.getTableInfoPrecursor(), "ModifiedSequence", new SQLFragment("prec.ModifiedSequence"),true));
+        result.addAll(getConflictedPrecursors(container, TargetedMSManager.getTableInfoMolecule(), TargetedMSManager.getTableInfoMoleculePrecursor(), "IonFormula", new SQLFragment("COALESCE(prec.CustomIonName, prec.IonFormula)"), false));
```
Prefer prec.IonFormula over prec.CustomIonName in the COALESCE since precursors of a molecule will have the same CustomIonName.  Both IonFormula and CustomIonName are nullable, so maybe add GeneralPrecursor.mz here as it is done in molExprCol in MoleculeTableInfo?

labkey-jeckels replied:
Changed this to concatenate the values, plus mz

vagisha commented on src/org/labkey/targetedms/query/ConflictResultsManager.java:
```diff
@@ -88,67 +87,76 @@ private ConflictResultsManager() {}
+        sql.append(precursorTable, "prec2");
+        sql.append(" ON prec." + joinColumn + " = prec2." + joinColumn);
+        sql.append(" INNER JOIN targetedms.generalprecursor gp2 ");
+        sql.append("ON (gp2.Id = prec2.Id AND gp.Charge = gp2.Charge) ");
```
I think we will need to add 'gp.mz = gp2.mz' in the join condition here.  The joinColumn ("IonFormula") in line 133 can be null.

labkey-jeckels replied:
Done

vagisha reviewed (approved):
This was a lot of work!! Thank you so much for doing this.   I left a few comments. 
I will give this a try later this week, hopefully.

Commit:
Incorporate code review feedback

Correct matching for small molecule precursors, using ion formula, name, and mz

Handle restoring previous precursors to the library when a run is deleted

Update test to cover deletion scenario

labkey-jeckels commented:
Thanks Vagisha. I addressed your concerns, which also led to to find some other problems, primarily related to how we match precursors with each other for small molecules.

Commit:
Fix experiment test

Improve small molecule detail pages

Commit:
Update expected label

Commit:
Fix typo

Commit:
Fix MoleculePrecursor ordering for CLIB columns

Commit:
More test updates

UI and navigation improvements for details pages
